### PR TITLE
Feat/MET-2018_MET-4364 error report mechanism dereference enrichment

### PR DESCRIPTION
**Pull request by jeortizquan**

### CODE REVIEW

**Review by stzanakis — COMMENTED** [review]
Seems okay overall.
There are some improvements or considerations though that we could apply.
Please have a look.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/Dereferencer.java [hunk 1]
```diff
@@ -14,17 +17,17 @@ public interface Dereferencer {
   /**
    * Performs the dereference procedure on a RDF
    * @param rdf The RDF to be processed
-   * @throws DereferenceException In case there is a problem with dereferencing.
+   * @return A report containing messages of dereference process
    */
-  void dereference(final RDF rdf) throws DereferenceException;
+  HashSet<ReportMessage> dereference(final RDF rdf);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/Dereferencer.java:22]: Updated to `Set` interface.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/report/ProcessedResult.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ */
+public class ProcessedResult<T> {
+
+  private final T processedRecord;
+  private HashSet<ReportMessage> reportMessages;
+  private final RecordStatus recordStatus;
+
+  /**
+   * Constructor with record
+   * @param processedRecord
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/report/ProcessedResult.java:19]: Updated documentation.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/report/ReportMessage.java [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+import java.util.Objects;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * ReportMessage class used to inform processed results
+ */
+public final class ReportMessage {
+
+  private int status;
+  private Mode mode;
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/report/ReportMessage.java:13]: done.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java [hunk 4]
```diff
@@ -42,95 +52,172 @@ public class DereferencerImpl implements Dereferencer {
...
-        // TODO: 25/01/2021 If the id is an invalid url we bypass it. In the future we might want to check each individual link and/or validate it
-        referenceTerm = new ReferenceTermImpl(new URL(id), new HashSet<>());
-        referenceTermSet.add(referenceTerm);
-      } catch (MalformedURLException e) {
-        LOGGER.debug("Invalid enrichment reference found: {}", id);
-      }
-    }
+    Set<ReferenceTerm> referenceTermSet = resourceIds
+        .stream()
+        .map(id -> {
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java]: extracted functions 1st phase.

File: metis-enrichment/metis-enrichment-client/src/main/resources/enrichment-sequence-diagram.txt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+sequenceDiagram
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/resources/enrichment-sequence-diagram.txt:1]: removed.

File: metis-enrichment/metis-enrichment-client/pom.xml [hunk 6]
```diff
@@ -23,6 +27,12 @@
       <groupId>org.mockito</groupId>
       <artifactId>mockito-core</artifactId>
     </dependency>
+    <dependency>
+      <groupId>com.github.tomakehurst</groupId>
+      <artifactId>wiremock-jre8</artifactId>
+      <version>2.34.0</version>
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/pom.xml:33]: added.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/model/EnrichmentResultBaseWrapper.java [hunk 7]
```diff
@@ -28,16 +29,20 @@ public class EnrichmentResultBaseWrapper {
       @XmlElement(name = "TimeSpan", namespace = "http://www.europeana.eu/schemas/edm/", type = TimeSpan.class)})
   private List<EnrichmentBase> enrichmentBase = new ArrayList<>();
 
+  private DereferenceResultStatus enrichmentStatus;
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/model/EnrichmentResultBaseWrapper.java:32]: :sweat_smile: I forgot to rename it. done.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/report/ReportMessage.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(status, mode, messageType, value, message, StringUtils.substring(stackTrace, 1, MAX_COMPARE_STACK_TRACE));
+  }
+
+  /**
+   * {@code ReportMessage} builder static inner class.
+   */
+  public static final class ReportMessageBuilder {
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/report/ReportMessage.java:81]: Did some updates to the class taking your suggestions, see if they are better now.

File: metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java [hunk 9]
```diff
@@ -107,56 +115,86 @@ public List<EnrichmentBase> dereference(String resourceId)
...
-    // Get the main object to dereference. If null, we are done.
-    final Pair<EnrichmentBase, Vocabulary> resource = computeEnrichmentBaseVocabularyPair(
-        resourceId);
-    if (resource == null) {
-      return Collections.emptyList();
-    }
+  private Pair<Collection<EnrichmentBase>, DereferenceResultStatus> dereferenceResource(String resourceId) {
+    try {
+      // Get the main object to dereference. If null, we are done.
+      final Triple<EnrichmentBase, Vocabulary, DereferenceResultStatus> resource = computeEnrichmentBaseVocabularyTriple(resourceId);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java]: Created result object to handle this.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java [hunk 10]
```diff
@@ -42,95 +52,172 @@ public class DereferencerImpl implements Dereferencer {
...
     LOGGER.debug("Dereference completed.");
-
+    return reportMessages;
   }
 
   @Override
-  public List<EnrichmentBase> dereferenceEntities(Set<String> resourceIds)
-      throws DereferenceException {
-
+  public Pair<List<EnrichmentBase>, HashSet<ReportMessage>> dereferenceEntities(Set<String> resourceIds) {
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java]: Created an object to handle the pair and triple.

File: metis-enrichment/metis-enrichment-client/src/test/resources/enriched.xml [hunk 11]
```diff
@@ -0,0 +1,460 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/test/resources/enriched.xml:1]: removed.

File: metis-enrichment/metis-enrichment-client/src/test/java/eu/europeana/enrichment/rest/client/EnrichmentWorkerImplTest.java [hunk 12]
```diff
@@ -11,21 +11,87 @@
...
+
+  @ParameterizedTest
+  @MethodSource("providedInputRecords")
+  void testEnrichmentWorkerHappyFlow(String inputRecord, RecordStatus recordStatus)
+      throws DereferenceException, EnrichmentException {
+    TreeSet<Mode> modeSetWithBoth = new TreeSet<>();
+    modeSetWithBoth.add(Mode.ENRICHMENT);
+    modeSetWithBoth.add(Mode.DEREFERENCE);
+    EnricherProvider enricherProvider = new EnricherProvider();
+    enricherProvider.setEnrichmentPropertiesValues("https://entity-management-production.eanadev.org/entity",
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/test/java/eu/europeana/enrichment/rest/client/EnrichmentWorkerImplTest.java]: Added the mocks.

File: metis-enrichment/metis-enrichment-rest/src/main/java/eu/europeana/enrichment/rest/EnrichmentController.java [hunk 13]
```diff
@@ -104,7 +109,7 @@ public EnrichmentBase equivalence(@ApiParam("uri") @RequestParam("uri") String u
   public EnrichmentResultList equivalence(@RequestBody EnrichmentReference enrichmentReference) {
     final List<EnrichmentResultBaseWrapper> enrichmentBaseWrappers = enrichmentReference
         .getReferenceValues().stream().map(enrichmentService::enrichByEquivalenceValues)
-        .map(EnrichmentResultBaseWrapper::new).collect(Collectors.toList());
+        .map(item -> new EnrichmentResultBaseWrapper(item, DereferenceResultStatus.SUCCESS)).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-rest/src/main/java/eu/europeana/enrichment/rest/EnrichmentController.java]: done.

File: metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/EnrichmentService.java [hunk 14]
```diff
@@ -1,5 +1,6 @@
 package eu.europeana.enrichment.service;
 
+import eu.europeana.enrichment.api.external.DereferenceResultStatus;
```
Comment (stzanakis) [inline metis-enrichment/metis-enrichment-service/src/main/java/eu/europeana/enrichment/service/EnrichmentService.java:3]: Not used

**Review by stzanakis — CHANGES_REQUESTED** [review]

File: metis-dereference/metis-dereference-common/pom.xml [hunk 15]
```diff
@@ -48,5 +48,11 @@
       <groupId>org.junit.jupiter</groupId>
       <artifactId>junit-jupiter-engine</artifactId>
     </dependency>
+    <dependency>
+      <groupId>eu.europeana.metis</groupId>
+      <artifactId>metis-enrichment-common</artifactId>
+      <version>${project.version}</version>
+      <scope>compile</scope>
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-dereference/metis-dereference-common/pom.xml:55]: grouped and moved at the start where all of them they are.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/model/EnrichmentResultBaseWrapper.java [hunk 16]
```diff
@@ -50,11 +68,19 @@ public List<EnrichmentBase> getEnrichmentBaseList() {
...
    * @return the converted list
    */
   public static List<EnrichmentResultBaseWrapper> createEnrichmentResultBaseWrapperList(
-      Collection<List<EnrichmentBase>> resultList) {
-    return resultList.stream().map(EnrichmentResultBaseWrapper::new).collect(Collectors.toList());
+      Collection<List<EnrichmentBase>> resultList, DereferenceResultStatus dereferenceStatus) {
+    return resultList.stream().map( item -> new EnrichmentResultBaseWrapper(item, dereferenceStatus)).collect(Collectors.toList());
   }
 
+  public void setDereferenceStatus(DereferenceResultStatus dereferenceStatus) {
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/api/external/model/EnrichmentResultBaseWrapper.java:79]: removed unused code.

File: metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java [hunk 17]
```diff
@@ -107,56 +111,104 @@ public List<EnrichmentBase> dereference(String resourceId)
    * </p>
    *
    * @param resourceId The resource to dereference.
-   * @return A collection of dereferenced resources. Is not null, but could be empty.
+   * @return A collection of Pair dereferenced resources and the status of dereference process. Is not null, but could be empty.
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java:114]: documentation improved.

File: metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java [hunk 18]
```diff
@@ -315,26 +386,33 @@ private String retrieveOriginalEntity(String resourceId, VocabularyCandidates ca
...
   }
 
-  Pair<EnrichmentBase, Vocabulary> computeEnrichmentBaseVocabularyPair(String resourceId)
+  EnrichmentEntityVocabulary computeEnrichmentBaseVocabularyTriple(String resourceId)
       throws JAXBException, URISyntaxException {
 
     // Try to get the entity and its vocabulary from the cache.
     final ProcessedEntity cachedEntity = processedEntityDao.getByResourceId(resourceId);
-    final Pair<String, Vocabulary> entityVocabularyPair = computeEntityVocabularyPair(resourceId,
+    final EnrichmentEntityVocabulary entityVocabularyTriple = computeEntityVocabularyTriple(resourceId,
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java:399]: It becomes simpler :smile:

File: metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java [hunk 19]
```diff
@@ -107,56 +111,104 @@ public List<EnrichmentBase> dereference(String resourceId)
...
-    if (resource == null) {
-      return Collections.emptyList();
+  private DereferenceResult dereferenceResource(String resourceId) {
+    DereferenceResult dereferenceResult = null;
+    try {
+      // Get the main object to dereference. If null, we are done.
+      //Triple<EnrichmentBase, Vocabulary, DereferenceResultStatus>
+      final EnrichmentEntityVocabulary resource = computeEnrichmentBaseVocabularyTriple(resourceId);
+
+      dereferenceResult = checkEmptyEnrichmentBaseAndVocabulary(dereferenceResult, resource);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java:127]: Yes. thanks for noticing, It is an error of refactoring the code, I just removed the extra parameter and placed the variable where I needed.

File: metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java [hunk 20]
```diff
@@ -186,14 +238,14 @@ private static <T> Stream<T> getStream(Collection<T> collection) {
    *
    * @param resourceId the url of the provider entity
    * @param cachedEntity the cached entity object
-   * @return a pair with the computed values
+   * @return a EnrichmentEntityVocabulary with the entity, vocabulary, and status.
    * @throws URISyntaxException if the resource identifier url is invalid
    * @throws TransformerException if an exception occurred during transformation of the original entity
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java:243]: Thanks, actually was not being catched the proper way, updated this section.

File: metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java [hunk 21]
```diff
@@ -107,56 +111,104 @@ public List<EnrichmentBase> dereference(String resourceId)
...
+        // Done
+        dereferenceResult = new DereferenceResult(
+            result.values().stream().map(Pair::getLeft).collect(Collectors.toList()),
+            result.values().stream().map(Pair::getRight).filter(Objects::nonNull).findFirst()
+                  .orElse(DereferenceResultStatus.UNKNOWN_ENTITY));
+      }
+    } catch (JAXBException jaxbException) {
+      LOGGER.warn(String.format("Problem occurred while dereferencing resource %s.", resourceId), jaxbException);
+      // No EnrichmentBase + Status
+      dereferenceResult = new DereferenceResult(Collections.emptyList(), DereferenceResultStatus.ENTITY_FOUND_XML_XLT_ERROR);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java:155]: refactored. done.

File: metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java [hunk 22]
```diff
@@ -107,56 +111,104 @@ public List<EnrichmentBase> dereference(String resourceId)
...
     }
+    return dereferenceResult;
+  }
 
-    // Create value resolver that catches exceptions and logs them.
-    final Function<String, EnrichmentBase> valueResolver = key -> {
-      Pair<EnrichmentBase, Vocabulary> result;
+  private Function<String, Pair<EnrichmentBase, DereferenceResultStatus>> getValueResolver() {
+    return key -> {
+      //Triple<EnrichmentBase, Vocabulary, DereferenceResultStatus>
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java:166]: removed. done.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java [hunk 23]
```diff
@@ -85,33 +99,128 @@ public void enrichment(RDF rdf) throws EnrichmentException {
...
+                                             .collect(Collectors.toList());
+    for (HttpStatus status : warningStatuses) {
+      if (message.contains(status.toString())) {
+        return status;
+      }
+    }
+    return null;
+  }
+
+  private static Throwable findRootCause(Throwable throwable) {
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:217]: Refactored to use apache.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java [hunk 24]
```diff
@@ -85,33 +99,128 @@ public void enrichment(RDF rdf) throws EnrichmentException {
...
+            .withValue(referenceValue)
+            .withException(rootCause)
+            .build());
+      }
+      return new ImmutablePair<>(null, reportMessages);
+    }
+  }
+
+  private static HttpStatus containsWarningStatus(String message) {
+    List<HttpStatus> warningStatuses = Arrays.stream(HttpStatus.values())
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/enrichment/EnricherImpl.java:205]: Excellent. Moved to a static field.

File: metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/EnrichmentEntityVocabulary.java [hunk 25]
```diff
@@ -0,0 +1,112 @@
+package eu.europeana.metis.dereference.service;
+
+import eu.europeana.enrichment.api.external.DereferenceResultStatus;
+import eu.europeana.enrichment.api.external.model.EnrichmentBase;
+import eu.europeana.metis.dereference.Vocabulary;
+
+/**
+ * Result of enrichment base or enrichment entity
+ */
+public class EnrichmentEntityVocabulary {
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/EnrichmentEntityVocabulary.java:10]: done.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/report/ReportMessage.java [hunk 26]
```diff
@@ -0,0 +1,213 @@
...
+import eu.europeana.enrichment.rest.client.EnrichmentWorker.Mode;
+import java.util.Objects;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.springframework.http.HttpStatus;
+
+/**
+ * ReportMessage class used to inform processed results
+ */
+public final class ReportMessage {
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/report/ReportMessage.java:12]: Renamed to report

File: metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java [hunk 27]
```diff
@@ -315,26 +386,33 @@ private String retrieveOriginalEntity(String resourceId, VocabularyCandidates ca
...
+    } else {
+      enrichmentBaseVocabulary = convertToEnrichmentBaseVocabulary(
+          entityVocabularyTriple.getEntity(), entityVocabularyTriple.getVocabulary());
+    }
+    if (enrichmentBaseVocabulary == null) {
+      return new EnrichmentEntityVocabulary(entityVocabularyTriple.getDereferenceResultStatus());
     } else {
-      enrichmentBaseVocabularyPair = convertToEnrichmentBaseVocabularyPair(
-          entityVocabularyPair.getLeft(), entityVocabularyPair.getRight());
+      return new EnrichmentEntityVocabulary(enrichmentBaseVocabulary.getEnrichmentBase(),
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-dereference/metis-dereference-service/src/main/java/eu/europeana/metis/dereference/service/MongoDereferenceService.java:413]: The code was entangled with processing similar conditions, I just refactored it and it became simpler, reusing a constructor.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/report/ReportMessage.java [hunk 28]
```diff
@@ -0,0 +1,213 @@
...
+    stackTrace = ExceptionUtils.getStackTrace(val);
+    return this;
+  }
+
+  /**
+   * Returns a {@code ErrorMessage} built from the parameters previously set.
+   *
+   * @return a {@code ErrorMessage} built with parameters of this {@code ErrorMessage.Builder}
+   */
+  public ReportMessage build() {
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/report/ReportMessage.java:210]: I simplified it to a single class due sonar reported as duplicated code.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/EnrichmentWorkerImpl.java [hunk 29]
```diff
@@ -52,75 +53,103 @@ public Set<Mode> getSupportedModes() {
...
-  public RDF process(final RDF rdf, Set<Mode> modes)
-      throws EnrichmentException, DereferenceException {
-
+  public ProcessedResult<RDF> process(final RDF rdf, Set<Mode> modes) {
+    HashSet<ReportMessage> reportMessages = new HashSet<>();
+    IllegalArgumentException e;
     // Sanity checks
     if (rdf == null) {
-      throw new IllegalArgumentException("Input RDF cannot be null.");
+      e = new IllegalArgumentException("Input RDF cannot be null.");
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/EnrichmentWorkerImpl.java:131]: Simplified it.

File: metis-enrichment/metis-enrichment-rest/src/main/java/eu/europeana/enrichment/rest/EnrichmentController.java [hunk 30]
```diff
@@ -61,7 +67,10 @@ public EnrichmentController(EnrichmentService enrichmentService) {
   public EnrichmentResultList search(
       @ApiParam("SearchTerms") @RequestBody EnrichmentSearch enrichmentSearch) {
     return new EnrichmentResultList(enrichmentService
-        .enrichByEnrichmentSearchValues(enrichmentSearch.getSearchValues()));
+        .enrichByEnrichmentSearchValues(enrichmentSearch.getSearchValues())
+        .stream()
+        .map(item -> new EnrichmentResultBaseWrapper(item.getEnrichmentBaseList(), DereferenceResultStatus.SUCCESS))
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply metis-enrichment/metis-enrichment-rest/src/main/java/eu/europeana/enrichment/rest/EnrichmentController.java:72]: I am defining it to `null`. in this class just for compatibility.

**Pull request merged**